include/deal.II/meshworker/scratch_data.h

```diff
@@ -0,0 +1,1301 @@
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_meshworker_scratch_data_h
+#define dealii_meshworker_scratch_data_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/differentiation/ad.h>
+
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/fe/mapping_q1.h>
+
+#include <boost/any.hpp>
+
+#include <algorithm>
+#include <map>
+#include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace MeshWorker
+{
+  namespace
+  {
+    /**
+     * Alias used to deduce the OutputType when asking for values or gradients
+     * of shape functions multiplied with the given @p NumberType.
+     */
+    template <int dim, int spacedim, typename NumberType, typename Extractor>
+    using OutputType = typename FEValuesViews::View<dim, spacedim, Extractor>::
+      template OutputType<NumberType>;
+  } // namespace
+
+  /**
+   * A helper class to simplify the parallel assembly of linear and non-linear
+   * problems, and the evaluation of finite element fields.
+   *
+   * This class is a drop in ScratchData to use with the WorkStream::run()
+   * function and with the MeshWorker::mesh_loop function().
+   *
+   * The ScratchData class has three main goals:
+   * - to create FEValues, FEFaceValues, and FESubFaceValues for the current
+   *   cell and for its neighbor cell *on demand* (only if they are
+   *   necessary for the algorithm provided by the user), and to provide a
+   *   uniform interface to access the FEValues objects when assembling cell,
+   *   face, or subface contributions
+   * - to store arbitrary data types (or references to arbitrary data types),
+   *   that can be retrieved by name
+   * - to provide a reasonable interface for those use cases where the user may
+   *   need temporary vectors of data at quadrature points, allowing the
+   *   construction of these temporaries *on demand*, and easy access to
+   *   values, gradients, etc. of already computed solution vectors.
+   *
+   * An example usage for this class is given by the following snippet of code:
+   *
+   * @code
+   * ScratchData<dim,spacedim> scratch(fe,
+   *                                   cell_quadrature,
+   *                                   update_values | update_gradients);
+   *
+   * FEValuesExtractors::Vector velocity(0);
+   * FEValuesExtractors::Scalar pressure(dim);
+   *
+   * ...
+   *
+   * for(const auto &cell : dof_handler.active_cell_iterators()) {
+   *    const auto &fe_values = scratch.reinit(cell);
+   *    const auto &local_solution_values =
+   *          scratch.extract_local_dof_values("previous_solution",
+   *                                           previous_solution);
```

Comment:
did you mean to reference `current_solution` here instead of `previous_solution`?